server/server.go

```diff
@@ -75,33 +78,30 @@ func (s *Server) RunUntilShutdown() error {
 	// the interrupt handler.
 	<-s.interceptor.ShutdownChannel()
 
+	// shutdown servers
+	s.grpcServer.GracefulStop()
+	s.httpServer.Shutdown(context.Background())
+
 	return nil
 }
 
 func (s *Server) startGrpcServer() error {
-	lis, err := net.Listen("tcp", s.cfg.GRPCListener)
+	listener, err := net.Listen("tcp", s.cfg.GRPCListener)
 	if err != nil {
 		return fmt.Errorf("failed to listen on %s: %w", s.cfg.GRPCListener, err)
 	}
-	defer lis.Close()
 
 	grpcServer := grpc.NewServer()
-	defer grpcServer.Stop()
-
 	proto.RegisterFinalityGadgetServer(grpcServer, s)
 
-	var wg sync.WaitGroup
-	wg.Add(1)
+	ready := make(chan struct{})
```

Comment:
why replace WaitGroup to use channel?

also why is the channel name `ready`?

Reply:
WaitGroup only signals goroutine started, not if server is ready to accept connections, "ready" is used b/c it is clear for the channel's purpose

Reply:
> WaitGroup only signals goroutine started, not if server is ready to accept connections

but how does it work in the past? we have always been using wg

> "ready" is used b/c it is clear for the channel's purpose

ask Cursor. it's not a good name

Reply:
> how does it work in the past?

right, WaitGroup works in server setups, but my understanding is the key difference is channels can propagate info while WaitGroup can only signal completion

> it's not a good name

yes, it is too generic, maybe `listenerReady` is better name